site/src/contexts/ProxyContext.tsx

```diff
@@ -6,25 +6,60 @@ import {
   createContext,
   FC,
   PropsWithChildren,
+  useCallback,
   useContext,
+  useEffect,
   useState,
 } from "react"
 import { ProxyLatencyReport, useProxyLatency } from "./useProxyLatency"
 
-interface ProxyContextValue {
+export interface ProxyContextValue {
+  // proxy is **always** the workspace proxy that should be used.
+  // The 'proxy.selectedProxy' field is the proxy being used and comes from either:
+  //   1. The user manually selected this proxy. (saved to local storage)
+  //   2. The default proxy auto selected because:
+  //    a. The user has not selected a proxy.
+  //    b. The user's selected proxy is not in the list of proxies.
+  //    c. The user's selected proxy is not healthy.
+  //   3. undefined if there are no proxies.
+  //
+  // The values 'proxy.preferredPathAppURL' and 'proxy.preferredWildcardHostname' can
+  // always be used even if 'proxy.selectedProxy' is undefined. These values are sourced from
+  // the 'selectedProxy', but default to relative paths if the 'selectedProxy' is undefined.
   proxy: PreferredProxy
+
+  // userProxy is always the proxy the user has selected. This value comes from local storage.
+  // The value `proxy` should always be used instead of `userProxy`. `userProxy` is only exposed
+  // so the caller can determine if the proxy being used is the user's selected proxy, or if it
+  // was auto selected based on some other criteria.
```

Comment:
Hmm, curious if we could  just check local storage when this check is needed instead of exposing it on the context value. Makes the API a bit confusing.

Reply:
We absolutely could, but I wanted the entire state to be exposed on this context.

If we expect the caller to check local storage, it could also be inferred the caller could mutate local storage on their own.

By encapsulating it this way, we can move away from local storage to some other storage mechanism if we decide to later make this some user config value stored in the database.

---

In summary, for encapsulation reasons, I prefer the context to control this.